### ISSUE TITLE
External change message shows when making multiple saves quickly

### ISSUE THREAD

**Issue by vahid-sanati** [adobe/brackets#12867]
When making multiple saves quickly after each other, this message shows up. It doesn't happen if there is more than 2sec delay between two saves but it's showing when making saves quickly. this started since #12865 after applying line ending normalization.

![cc089596-9fb3-11e6-9a80-4781db5ccbbe](https://cloud.githubusercontent.com/assets/19893777/19885609/ce28c752-a034-11e6-8217-544370e749b9.png)

**Comment by petetnt** [comment]
How many lines is `control.php`? I am bit worried about the O(n) nature of the RegExp replace in #12865.

**Comment by vahid-sanati** [comment]
@petetnt it's just 63 lines.

**Comment by petetnt** [comment]
Well that's very weird 🤔

**Comment by zaggino** [comment]
Can't replicate this on my (quite powerful) machine.

@vahid-sanati can you confirm that if you remove the fix in https://github.com/adobe/brackets/pull/12865, it stops?

It could be happening due to relatively slow operating system reporting file system changes in wrong order. You're running Windows, right @vahid-sanati ?

**Comment by vahid-sanati** [comment]
@zaggino yes I'm running windows 10 x64. That's weird! the message showed up even without the #12865 fix. I just noticed this issue after that fix and think that was the reason.

I think the event that listens to file change must be paused while brackets is saving the file and continue after save is completed.

**Comment by vahid-sanati** [comment]
when the message shows, if I select to reload from disk, the last change reverts back. maybe this give a clue to you.

**Comment by zaggino** [comment]
I believe it's basically this scenario:
1) you save a change A
2) FS notifies the file has been changed
3) Brackets read the file containing A
4) you save a change B
5) Brackets evaluate that A is different to B
The fix would be something of ignoring 5) if a save was done after 3)

**Comment by vahid-sanati** [comment]
sometimes it gets fixed by itself and never happens no matter how quick I save. and sometimes it starts happening again. but the strange part is when this issue was fixed the undo issue was fixed too. (using the one without #12865 fix). it has become completely random.

**Comment by vahid-sanati** [comment]
found the situation that the problem is happening and not happening. when opening the file for first time or rename the project the problem is not happening for first time. after closing file and brackets and reopening the file, the problem begins happening. the same situation I described in undo issue before discovering new line.

**Comment by zaggino** [comment]
Since a certain speed is required to reproduce this, it'll probably be easier to replicate with an integration test-case. I'd vote this not to be a blocker for releasing 1.8 though, as it's not a new issue and honestly, not many people save a file every 1-2 seconds.

**Comment by vahid-sanati** [comment]
@zaggino I don't save every 1-2 seconds either :D I agree with you this issue is not that important to block 1.8 release. I just saw this and wanted to show you that maybe it can give you some clue because I believe this is related to undo reset problem. The reason is as I said before, in the situations that this problem doesn't show up, the undo is not resetting too. (without using the fixed version)

**Comment by vahid-sanati** [comment]
This maybe useful to mention. every time I hit save, the folding arrows flash once. before undo fix that meant I lost undo. but after fix the undo is not resetting anymore but the flash still happens. this means the editor assumes the changes to be external and reads the file from disk but that text comparison you made prevents resetting history. It's not properly detecting which change is internal and which one is external.

**Comment by swmitra** [comment]
I doesn't look like to be related to #12865 changes. It looks like a timing issue which occurs on a very specific sequence and dependent on io speed of the machine. 
At first glance I guess the scenario must be like - 
* Save a file 
* Brackets gets file modified notification 
* Does something in main thread and triggers an io to read content
* Now comes the second notification 
* By this time the doc content in memory is updated as the edit took place in Brackets itself
* Again we do something and trigger an io

Now if we get the callback from the first io request now, the text comparison will fail and we will detect this as a file content change which has happened externally. As this callback order is directly dependent on io and processing speed we might see the reported behavior on some machines. Not sure if my assumption is correct, but if it's true then we need to spend some time to fix this either by having an in-memory doc content snapshot which we compare with the io callback content, or to employ something which should discard the callback handler if we have spawned another io request.
